README.md
English | [简体中文](README-CN.md)

![](https://aliyunsdk-pages.alicdn.com/icons/AlibabaCloud.svg)

## Alibaba Cloud Darabonba Time

## Installation

- [Go](./golang/README.md)
- [.Net](./csharp/README.md)

## Issues

[Opening an Issue](https://github.com/aliyun/darabonba-time/issues/new), Issues not conforming to the guidelines may be closed immediately.

## References

- [Latest Release](https://github.com/aliyun/darabonba-time)

## License

[Apache-2.0](http://www.apache.org/licenses/LICENSE-2.0)

Copyright (c) 2009-present, Alibaba Cloud All rights reserved.


csharp/ChangeLog.md
### 2021-01-05 Version 0.0.1
* First release.

php/tests/TimeTest.php
<?php

namespace AlibabaCloud\Darabonba\Time\Tests;

use AlibabaCloud\Darabonba\Time\Time;
use PHPUnit\Framework\TestCase;

/**
 * @internal
 * @coversNothing
 */
class TimeTest extends TestCase
{
    public function testUnix()
    {
        $time = Time::unix(); // is string
        $this->assertFalse($time === (int) $time);
        $this->assertTrue(abs(time() - (int) $time) <= 3);
    }

    public function testUTC()
    {
        $t = sprintf('%.6f', strtotime('2020-02-29 23:11:48'));
        $t = bcadd($t, 0.123456, 6);
        $this->assertEquals('2020-02-29 23:11:48.123456 +0000 UTC', Time::UTC($t));
    }

    public function testSleep()
    {
        $start = microtime(true) * 1000;
        Time::sleep(200);
        $end = microtime(true) * 1000;
        $this->assertTrue($end - $start >= 200);
    }

    public function testFormat()
    {
        $res = Time::format('yyyyMMdd hh:mm:ss a EEEE', strtotime('2020-02-29 23:11:48'));
        $this->assertEquals('20200229 11:11:48 pm Saturday', $res);
    }
}


ts/test/client.spec.ts
import Client from '../src/client';
import assert from 'assert';
import moment from 'moment';
import 'mocha';

describe('Darabonba Time', function () {
  it('Module should ok', function () {
    assert.ok(Client);
  });

  it('unix should ok', function () {
    assert.deepStrictEqual(Client.unix().length, 10);
  });

  it('UTC should ok', function () {
    assert.ok(Client.UTC().endsWith('UTC'));
  });

  it('format should ok', function () {
    assert.deepStrictEqual(Client.format('yyyyMMdd hh:mm a EEEE'), moment().format('YYYYMMDD HH:mm A dddd'));
  });
});

csharp/tests/TimeTest.cs
using System;
using AlibabaCloud.DarabonbaTime;
using Xunit;

namespace tests
{
    public class TimeTest
    {
        [Fact]
        public void Test_Unix()
        {
            Assert.Equal(10, Time.Unix().Length);
        }

        [Fact]
        public void Test_UTC()
        {
            Assert.Contains("+0000 UTC", Time.UTC());
        }

        [Fact]
        public void Test_Format()
        {
            Assert.NotNull(Time.Format("yyyy-MM-dd"));
            Assert.Equal(10, Time.Format("yyyy-MM-dd").Length);
        }
    }
}


ts/src/client.ts
// This file is auto-generated, don't edit it
/**
 * This is a time module
 */
import * as $tea from '@alicloud/tea-typescript';
import moment from 'moment';


export default class Client {

  /**
   * Return the time stamp string,unit second
   * @return the time string, e.g 1491888244
   */
  static unix(): string {
    return moment().unix().toString();
  }

  /**
   * Return the UTC time string now
   * @return the time string, e.g 2020-11-24 09:44:45.426237 +0000 UTC
   */
  static UTC(): string {
    return moment().format('YYYY-MM-DD HH:mm:ss.SSSSSS ZZ UTC');
  }

  /**
   * Suspends the current thread for the specified number of milliseconds.
   */
  static async sleep(millisecond: number): Promise<void> {
    return new Promise(resolve => {
      setTimeout(() => {
        resolve();
      }, millisecond)
    })
  }

  /**
   * Format the UTC time now with the layout
   * @param layout the format of time, e.g yyyyMMdd hh:mm a EEEE
   * @return the time string, e.g 20060102 15:04 PM Monday
   */
  static format(layout: string): string {
    layout = layout.replace(/y/g, 'Y')
      .replace(/d/g, 'D').replace(/h/g, 'H')
      .replace(/a/g, 'A').replace(/E/g, 'd');
    return moment().format(layout);
  }

}


csharp/core/Time.cs
/**
 * This is a time module
 */
// This file is auto-generated, don't edit it. Thanks.

using System;
using System.Threading;


namespace AlibabaCloud.DarabonbaTime
{
    public class Time 
    {
        private static readonly DateTime Jan1st1970 = new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);
        
        /**
         * Return the time stamp string,unit second
         * @return the time string, e.g 1491888244
         */
        public static string Unix()
        {
            return ((long)(DateTime.UtcNow - Jan1st1970).TotalSeconds).ToString();
        }

        /**
         * Return the UTC time string now
         * @return the time string, e.g 2020-11-24 09:44:45.426237 +0000 UTC
         */
        public static string UTC()
        {
            return DateTime.UtcNow.ToString("yyyy-MM-dd HH:mm:ss.ffffff '+0000 UTC'");
        }

        /**
         * Suspends the current thread for the specified number of milliseconds.
         */
        public static void Sleep(int t)
        {
            Thread.Sleep(t);
        }

        /**
         * Format the UTC time now with the layout
         * @param layout the format of time, e.g yyyyMMdd hh:mm a EEEE
         * @return the time string, e.g 20060102 15:04 PM Monday
         */
        public static string Format(string layout)
        {
            return DateTime.UtcNow.ToString(layout);
        }

    }
}


golang/client/client.go
// This file is auto-generated, don't edit it. Thanks.
/**
 * This is a time module
 */
package client

import (
	"fmt"
	"strings"
	"time"

	"github.com/alibabacloud-go/tea/tea"
)

func Unix() (_result *string) {
	return tea.String(fmt.Sprintf("%d", time.Now().UTC().Unix()))
}

func UTC() (_result *string) {
	return tea.String(time.Now().UTC().String())
}

func Sleep(t *int64) {
	time.Sleep(time.Duration(tea.Int64Value(t)))
}

func Format(layout *string) (_result *string) {
	l := tea.StringValue(layout)
	l = strings.Replace(l, "yyyy", "2006", 1)
	l = strings.Replace(l, "MM", "01", 1)
	l = strings.Replace(l, "dd", "02", 1)
	l = strings.Replace(l, "hh", "15", 1)
	l = strings.Replace(l, "mm", "04", 1)
	l = strings.Replace(l, "ss", "05", 1)
	l = strings.Replace(l, "a", "PM", 1)
	l = strings.Replace(l, "EEEE", "Monday", 1)
	l = strings.Replace(l, "E", "Mon", 1)
	return tea.String(time.Now().UTC().Format(l))
}


csharp/core/Properties/AssemblyInfo.cs
using System.Reflection;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;

// General Information about an assembly is controlled through the following 
// set of attributes. Change these attribute values to modify the information
// associated with an assembly.
[assembly: AssemblyTitle("")]
[assembly: AssemblyDescription("")]
[assembly: AssemblyConfiguration("")]
[assembly: AssemblyCompany("")]
[assembly: AssemblyProduct("")]
[assembly: AssemblyCopyright("")]
[assembly: AssemblyTrademark("")]
[assembly: AssemblyCulture("")]

// Setting ComVisible to false makes the types in this assembly not visible 
// to COM components.  If you need to access a type in this assembly from 
// COM, set the ComVisible attribute to true on that type.
[assembly: ComVisible(false)]

// The following GUID is for the ID of the typelib if this project is exposed to COM
[assembly: Guid("8c400c60-32b1-11eb-99fa-5d5d81892cd5")]

// Version information for an assembly consists of the following four values:
//
//      Major Version
//      Minor Version 
//      Build Number
//      Revision
//
[assembly: AssemblyVersion("1.0.0.0")]
[assembly: AssemblyFileVersion("1.0.0.0")]


php/src/Time.php
<?php

// This file is auto-generated, don't edit it. Thanks.

namespace AlibabaCloud\Darabonba\Time;

/**
 * This is a time module.
 */
class Time
{
    /**
     * Return the time stamp string,unit second.
     *
     * @return string the time string, e.g 1491888244
     */
    public static function unix()
    {
        return (string)time();
    }

    /**
     * Return the UTC time string now.
     *
     * @param int $time unix second
     *
     * @return string the time string, e.g 2020-11-24 09:44:45.426237 +0000 UTC
     */
    public static function UTC($time = null)
    {
        if (null === $time) {
            $m = substr((string) microtime(), 1, 7);

            return gmdate("Y-m-d H:i:s{$m} +0000 \\U\\T\\C", $time);
        }
        $m = bcsub($time, (int) $time, 6);
        $m = substr((string) $m, 1, 7);

        return gmdate("Y-m-d H:i:s{$m} +0000 \\U\\T\\C", $time);
    }

    /**
     * Suspends the current thread for the specified number of milliseconds.
     *
     * @param int $t
     */
    public static function sleep($t)
    {
        usleep($t * 1000);
    }

    /**
     * Format the UTC time now with the layout.
     *
     * @param string $layout the format of time, e.g yyyyMMdd hh:mm a EEEE
     * @param int    $time   unix second
     *
     * @return string the time string, e.g 20060102 15:04 PM Monday
     */
    public static function format($layout, $time = null)
    {
        $map  = [
            'yyyy' => 'Y',
            'MM'   => 'm',
            'dd'   => 'd',
            'HH'   => 'H',
            'hh'   => 'h',
            'mm'   => 'i',
            'ss'   => 's',
            'EEEE' => '%\w%',
        ];
        $week = [
            'Sunday',
            'Monday',
            'Tuesday',
            'Wednesday',
            'Thursday',
            'Friday',
            'Saturday',
        ];
        $w    = $week[(int) date('w', $time)];
        foreach ($map as $s => $t) {
            $layout = str_replace($s, $t, $layout);
        }
        $d = gmdate($layout, $time);

        return str_replace('%w%', $w, $d);
    }
}
